transaction/src/manifest/token.rs

```diff
@@ -0,0 +1,107 @@
+use sbor::rust::fmt;
+use sbor::rust::fmt::Debug;
+
+/// The span of tokens. The `start` and `end` are Unicode code points / UTF-32 - as opposed to a
+/// byte-based / UTF-8 index.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Span {
+    /// The start of the span, exclusive
+    pub start: Position,
+    /// The end of the span, inclusive
+    pub end: Position,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Position {
+    /// A 0-indexed cursor indicating the next unicode char from the start
+    pub full_index: usize,
+    /// A 1-indexed cursor indicating the line number (assuming \n is a line break)
+    pub line_number: usize,
+    /// A 0-indexed cursor indicating the character offset in the line
+    pub line_char_index: usize,
+}
+
+#[macro_export]
+macro_rules! position {
+    ($full_index:expr, $line_number:expr, $line_char_index:expr) => {
+        Position {
+            full_index: $full_index,
+            line_number: $line_number,
+            line_char_index: $line_char_index,
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! span {
+    (start = ($st_full_index:expr, $st_line_number:expr, $st_line_char_index:expr),
+         end = ($end_full_index:expr, $end_line_number:expr, $end_line_char_index:expr)) => {
+        Span {
+            start: position!($st_full_index, $st_line_number, $st_line_char_index),
+            end: position!($end_full_index, $end_line_number, $end_line_char_index),
+        }
+    };
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TokenKind {
+    // ==============
+    // Literals
+    // ==============
+    BoolLiteral(bool),
+    I8Literal(i8),
+    I16Literal(i16),
+    I32Literal(i32),
+    I64Literal(i64),
+    I128Literal(i128),
+    U8Literal(u8),
+    U16Literal(u16),
+    U32Literal(u32),
+    U64Literal(u64),
+    U128Literal(u128),
+    StringLiteral(String),
+
+    Ident(String),
+
+    /* Punctuations */
+    OpenParenthesis,
+    CloseParenthesis,
+    LessThan,
+    GreaterThan,
+    Comma,
+    Semicolon,
+    FatArrow,
+}
+
+impl fmt::Display for TokenKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            TokenKind::BoolLiteral(value) => write!(f, "value '{:?}' of Bool type", value),
```

Comment:
According to SBOR semantics, `type` is something else, really these are values, with e.g. Bool value kind.

Although _actually_ for the lexer, these are actually literals not values yet.

Perhaps we should just display the literal if it's a literal - e.g.:
* bool `"{:?}"` - i.e. `true` / `false`
* string: `"\"{}\""` - i.e. `"hello world"`
* integer: `"{:?}i8"` - i.e. `12u8`

And then for idents, say `"ident '{}'"` to keep the capitalization consistent with tokens?

Reply:
Done